Add layers to cluster graphs, improve some doc strings, expand tests

bannanc opened the pull request:
In this pull request I've add code to add atoms without smirks-indices to cluster graphs. 
It seems to be working the way I would want so far, but I've only tested a few cases, I would like to test more and update the relevant example notebook before we merge.

@davidlmobley or @jmaat if you have time to take a look that would be great.

davidlmobley commented on chemper/graphs/cluster_graph.py:
```diff
@@ -100,6 +105,36 @@ def add_atom(self, atom):
+            Returns
+            -------
+            score: int
+                 how similar is atom to current storage, max of 7 for all decorators identical
```
"How similar is atom to current storage" is opaque to me. Can you clarify? Does this mean basically out of the total number of possible decorators (7?) how many of these are the same? (And it is only applied when the element is already the same?)

bannanc replied:
Yes, your intuition was correct, but I will update the wording so that is more clear.

davidlmobley commented on chemper/graphs/cluster_graph.py:
```diff
@@ -100,6 +105,36 @@ def add_atom(self, atom):
+            -------
+            score: int
+                 how similar is atom to current storage, max of 7 for all decorators identical
+                 0 if atom's atomic number not included in current set
```
"0 if atom's atomic number..." also not totally clear. Does this mean that if you have a wildcard atom but with similar decorators it is still 0?

bannanc replied:
So I hadn't thought through the "wildcard" option completely. The input is always an atom, but its true that technically the atom storage (self) could be empty, which right now would cause a problem, I think it makes most sense to give that a high score since if the storage is empty it should match with any atom. I need to think about the best way to handle that scenario.

davidlmobley commented on chemper/graphs/cluster_graph.py:
```diff
@@ -100,6 +105,36 @@ def add_atom(self, atom):
+            """
+            score = 0
+            decs = self.make_atom_decorators(atom)
+            atomic_nums = [d[0] for d in self.decorators]
```
What is `d[0]` if it is a wildcard? I'm unclear what this does for wildcards.

davidlmobley commented on chemper/graphs/cluster_graph.py:
```diff
@@ -262,6 +297,128 @@ def _add_first_smirks_atoms(self, mol, smirks_atoms):
                                          neighbor_storage,
                                          bond=bond_storage)
 
+        for smirks_index, atom_index in smirks_atoms.items():
```
Seems like you should have comments here; not totally obvious what you're doing without thinking about it for a while/reading context.

bannanc replied:
I was going to add more detail to the new functions, but this one could probably use more help too.

davidlmobley commented on chemper/graphs/cluster_graph.py:
```diff
@@ -262,6 +297,128 @@ def _add_first_smirks_atoms(self, mol, smirks_atoms):
+            storage = self.atom_by_smirks_index[smirks_index]
+            self._add_layers(mol, atom, storage, self.layers, atom_dict)
+
+    def _add_layers(self, mol, atom, storage, layers, idx_dict):
```
This part is also probably involved enough you should have comments that briefly explain the algorithm to make the code easier to follow.

davidlmobley commented on chemper/graphs/fragment_graph.py:
```diff
@@ -152,13 +152,15 @@ def as_smirks(self):
 
         if self.atom_by_smirks_index:
-            min_smirks = min(self.atom_by_smirks_index)
+            # sometimes we use negative numbers for internal indexing
```
Why would negative indices be used?

bannanc replied:
I'm using negative indices to track the location of non-indexed atoms. ChemicalEnvironments do this internally too where alpha or beta atoms are located by their negative smirks index. I probably need to add more information in the doc strings on what those values are allowed to be.

bannanc replied:
I'm going to rename this as label, with details that positive labels are included in the SMIRKS patter, but that all other labels are just used as identification for storage.

davidlmobley commented on chemper/graphs/fragment_graph.py:
```diff
@@ -271,7 +273,7 @@ def add_atom(self, new_atom, new_bond=None, bond_to_atom=None,
         new_atom_storage = self.AtomStorage(new_atom, smirks_index=new_smirks_index)
         self._graph.add_node(new_atom_storage)
-        if new_smirks_index is not None and new_smirks_index > 0:
+        if new_smirks_index is not None:
```
You removed a check that only does this if the index is greater than 0; deliberate? Why? (Just confirming.)

bannanc replied:
yes, it was deliberate. Originally I was thinking of `atom_by_smirks_index` for only atoms that are actually tagged with smirks indices used in the smirks pattern (positive values), however, I realized for the cluster graph that using unique smirks indices made a convenient way to book keep atoms so I wanted the dictionary to have all of them.

davidlmobley commented on tests/test_basic_graphs.py:
```diff
@@ -0,0 +1,56 @@
+    assert c.as_smirks() == '[*:1]'
+
+
+# make sure corner cases can at least be built without failing
```
Why are these corner cases?

bannanc replied:
Corner cases was maybe not the best word for them. They were mostly added to increase coverage. Now that I think about it the `test_no_fail_fragment` could probably also have expected smirks patterns for at least the 1 layer options, so maybe those should go above. 

The multiple layers for the cluster graphs become really really complicated to write out and because decorators are stored as sets and not lists the SMIRKS patterns aren't consistent, I could maybe make a sorting option that makes the smirks patterns consistent so that more intense testing can be done instead of just "does it not fail"

davidlmobley commented on tests/test_graphs_output.py:
```diff
@@ -0,0 +1,95 @@
+def make_frag_graph(smiles, layers):
+    """
+    Generates a chemper Mol from the provided smiles
+    a ChemPerGraphFromMol with that molecules with the layers specified
```
The description of this test does not seem comprehensible.

bannanc commented:
I'm going to go ahead and merge this PR. 
I've addressed all the comments David brought up before. 
I didn't end up adding a new notebook example, but I'd like to finish updating the docs and current examples so anything I add is more clearly different.
